python/packaging/frontend_sdist/setup.py

```diff
@@ -15,42 +15,88 @@
 # limitations under the License.
 #
 
+import os
+import subprocess
 import sys
 
 from setuptools import setup
 from setuptools.command.install import install
-import subprocess as sp
 
 tensorrt_module = "##TENSORRT_MODULE##"
+tensorrt_version = "##TENSORRT_PYTHON_VERSION##"
+tensorrt_submodules = [
+    "{}_libs=={}".format(tensorrt_module, tensorrt_version),
+    "{}_bindings=={}".format(tensorrt_module, tensorrt_version),
+]
 
 
 class InstallCommand(install):
     def run(self):
-        def install_dep(package_name):
-            status = sp.run(
-                [
-                    sys.executable,
-                    "-m",
-                    "pip",
-                    "install",
-                    "{:}==##TENSORRT_PYTHON_VERSION##".format(package_name),
-                    "--index-url",
-                    "https://pypi.nvidia.com",
-                ]
-            )
-            status.check_returncode()
+        # pip-inside-pip hack ref #3080
+        subprocess.check_call(
+            [
+                "{}/bin/pip".format(sys.exec_prefix),
+                "install",
+                "--extra-index-url",
+                "https://pypi.nvidia.com",
+                *tensorrt_submodules,
+            ]
+        )
 
-        install_dep("{:}_libs".format(tensorrt_module))
-        install_dep("{:}_bindings".format(tensorrt_module))
+        super().run()
 
-        install.run(self)
+
+def parent_command_line():
+    """Get the command line of the parent PID."""
+    pid = os.getppid()
+    # try retrieval using psutil
+    try:
+        import psutil
+
+        return " ".join(psutil.Process(pid).cmdline())
+    except ModuleNotFoundError:
+        pass
+    # fall back to shell
+    try:
+        return (
+            subprocess.check_output(["ps", "-p", str(pid), "-o", "command"])
+            .decode()
+            .split("\n")[1]
+        )
+    except subprocess.CalledProcessError:
+        return ""
+
+
+# use pip-inside-pip hack only if the nvidia index is not set in the environment
+if (
+    "pypi.nvidia.com" in os.environ.get("PIP_EXTRA_INDEX_URL", "")
+    or "pypi.nvidia.com" in parent_command_line()
+):
+    install_requires = tensorrt_submodules
+    cmdclass = {}
+else:
+    install_requires = []
+    cmdclass = {"install": InstallCommand}
```

Comment:
I thought `install_requires` is encoded in the sdist metadata so this logic would only run when we build the package and not when a user installs it.

Reply:
you're right about the `METADATA` file in the bdist_wheel (it's leading and used for resolving, no more setup.py is run).

in the sdist there's a `PKG-INFO` file which will contain the deps (or not) depending on the logic in setup.py.

but that file is ignored by pip during install: when you only upload an sdist to pypi, pip will essentially untar into a temp dir, build a wheel on the local system, and install the resulting wheel.

so this approach will work as long as you continue to only upload the sdist:

```console
$ export PIP_EXTRA_INDEX_URL=https://pypi.nvidia.com
$ python setup.py sdist
$ pip install dist/tensorrt-0.8.2.tar.gz
Looking in indexes: https://pypi.org/simple, https://pypi.nvidia.com
Processing ./dist/tensorrt-0.8.2.tar.gz
  Preparing metadata (setup.py) ... done
INFO: pip is looking at multiple versions of tensorrt to determine which version is compatible with other requirements. This could take a while.
ERROR: Could not find a version that satisfies the requirement tensorrt_libs==0.8.2 (from tensorrt) (from versions: 8.6.1)
ERROR: No matching distribution found for tensorrt_libs==0.8.2
$ unset PIP_EXTRA_INDEX_URL
$ python setup.py sdist
$ pip install dist/tensorrt-0.8.2.tar.gz
Processing ./dist/tensorrt-0.8.2.tar.gz
  Preparing metadata (setup.py) ... done
Building wheels for collected packages: tensorrt
  Building wheel for tensorrt (setup.py) ... error
  error: subprocess-exited-with-error

  × python setup.py bdist_wheel did not run successfully.
  │ exit code: 1
  ╰─> [65 lines of output]
      ...
      installing to build/bdist.macosx-10.15-x86_64/wheel
      running install
      Looking in indexes: https://pypi.org/simple, https://pypi.nvidia.com
      ERROR: Could not find a version that satisfies the requirement tensorrt_libs==0.8.2 (from versions: 8.6.1)
      ERROR: No matching distribution found for tensorrt_libs==0.8.2
```

Reply:
Ah awesome, that should work for us then.